### PR TITLE
[FUSEQE-10617] Operatorhub automation

### PR DESCRIPTION
**Pull request by mmuzikar**

### CODE REVIEW

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Comment by mmuzikar** [comment]
@avano I am wondering whether we should also make something similar to -Pdeploy or add this logic to -Pdeploy. What do you think?

**Comment by avano** [comment]
> @avano I am wondering whether we should also make something similar to -Pdeploy or add this logic to -Pdeploy. What do you think?

:thinking: what would you want to deploy?

**Comment by mmuzikar** [comment]
Syndesis subscription on Operatorhub

**Comment by avano** [comment]
but that won't be used outside of the operatorhub tests, will it?

**Comment by mmuzikar** [comment]
yes that'd be just a convenience feature

**Comment by avano** [comment]
maybe it would be better to move the code for deploying from operatorhub into `Syndesis` class and then decide based on the properties set what deploy method to use?

like, if you have `syndesis.config.bundle.image` `syndesis.config.index.image` set -> use operator hub, otherwise use the standard method. wdyt?

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Comment by avano** [comment]
retest this please

**Comment by avano** [comment]
retest this please

**Comment by avano** [comment]
retest this please

**Comment by avano** [comment]
retest this please

**Comment by avano** [comment]
retest this please

**Comment by syndesisqe** [comment]
Following scenarios failed:
```
Integration - Database | Smoke - Periodic invocation to Insert 
```

**Comment by mkralik3** [comment]
Is step `Given deploy Syndesis via bundle` defined somewhere?  I am getting `The step "deploy Syndesis via bundle" is undefined.`

**Comment by mmuzikar** [comment]
@mkralik3 We decided to remove the scenario and rather use the deploy profile to setup operatorhub. I updated the README as well to make this clear.

**Comment by syndesisqe** [comment]
Following scenarios failed:
```
Integration - DB to DB | Read and update operations 
```

**Comment by avano** [comment]
retest this please

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

File: utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java [hunk 1]
```diff
@@ -705,4 +776,113 @@ public void linkServiceAccountWithSyndesisPullSecret(String serviceAccountName)
...
+            operatorDeployment.getJSONObject("spec").getJSONObject("template").getJSONObject("spec").getJSONArray("containers").getJSONObject(0)
+                .getJSONArray("env");
+        envVars.put(TestUtils.map("name", "TEST_SUPPORT", "value", "true"));
+        try {
+            OpenShiftUtils.getInstance().customResource(getCSVContext())
+                .edit(TestConfiguration.openShiftNamespace(), "fuse-online-operator.v7.8.0", json.toMap());
+        } catch (IOException e) {
+            log.error("Couldn't edit Syndesis CSV", e);
+        }
+    }
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java:887]: there is a code for that here: https://github.com/syndesisio/syndesis-qe/blob/master/utilities/src/main/java/io/syndesis/qe/TestSuiteParent.java#L37-L60

please merge it into one method that will do the following:
- TestSuiteParent will call that method
- in that method, check what type of deployment it is - either "standard" or operator hub
- for operator hub, check if the test support is already enabled, if no, enable it
- for standard deployment, check `if TestConfiguration.enableTestSupport()` and if yes, do that part of the code

File: utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java [hunk 2]
```diff
@@ -444,6 +480,41 @@ public void deployOperator() {
...
+            TestUtils.withRetry(() -> {
+                try {
+                    OpenShiftUtils.getInstance().updateDeploymentConfigEnvVars(operatorResourcesName, finalImagesEnvVars);
+                    return true;
+                } catch (KubernetesClientException kce) {
+                    log.debug("Caught KubernetesClientException: " + kce);
+                    return false;
+                }
+            }, 3, 30000L, "Unable to update operator deployment config after 3 tries");
+        }
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java:517]: :thinking: this part was changed in `1.11.x` so please adapt the changes for that

File: rest-tests/src/test/resources/features/operatorhub.feature [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    And create finish DB invoke sql action step with query "INSERT INTO TODO (task, completed) VALUES (:#task, 3)"
+    Then create integration with name: "DB to DB smoke rest test"
+    Then wait for integration with name: "DB to DB smoke rest test" to become active
+
+    And sleep for jenkins delay or 15 seconds
+
+    Then validate that number of all todos with task "Josef_first" is greater than 0
+    Then validate that number of all todos with task "Josef_second" is 0
+
```
Comment (avano) [inline rest-tests/src/test/resources/features/operatorhub.feature:31]: I don't know how, but there is a missing new line at the end :grin:

**Review by avano — CHANGES_REQUESTED** [review]

File: utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java (hunk 1 above)
Comment (mmuzikar) [reply utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java:887]: how does one merge? :smile: 
I think enabling test support needs to happen at different times for both ways. I need to edit the CSV after creating a subscription to ensure CSV is present on the cluster. And this code looks like it needs to run even prior to the operator existing.

File: utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java (hunk 2 above)
Comment (mmuzikar) [reply utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java:517]: I guess we could also remove this code and the feature file as the old operatorhub installation won't be used anymore, wdyt?

File: utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java (hunk 1 above)
Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java:887]: you're right, my bad :grin:

File: utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java (hunk 2 above)
Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/resource/impl/Syndesis.java:517]: makes sense

**Comment by syndesisqe** [comment]
Following scenarios failed:
```
Integration - DB to DB | Read and update operations 
```

**Review by avano — APPROVED** [review]

**Comment by avano** [comment]
retest this please

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Pull request merged**